### ISSUE TITLE
Invalid or unsupported protocol version

### ISSUE THREAD

**Issue by behzadsh** [gocql/gocql#1536]
Please answer these questions before submitting your issue. Thanks!

### What version of Cassandra are you using?
[cqlsh 5.0.1 | Cassandra 3.11.7 | CQL spec 3.4.4 | Native protocol v4]

### What version of Gocql are you using?
latest

### What version of Go are you using?
1.15.8

I am getting the following error every minute in the application logs:
```
error: failed to connect to x.x.x.x:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
```

Everything seems working fine but I get this error every minute and I don't know what is the problem. The error message is so confusing. 🤔

**Comment by martin-sucha** [comment]
It seems that the part `Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)` is from the server, do you see anything related in Cassandra server logs?

**Comment by behzadsh** [comment]
@martin-sucha no, I didn't notice any error or any message related to this.

**Comment by abhi93104** [comment]
Facing the same issue when trying to connect to cassandra cluster through gocql sdk

`gocql: unable to dial control conn 127.0.0.1:9042: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
panic: gocql: unable to create session: control: unable to connect to initial hosts: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)`

Code:
`func main() {

	cluster := gocql.NewCluster("127.0.0.1:9042")
	cluster.Authenticator = gocql.PasswordAuthenticator{
		Username: "username",
		Password: "pwd",
	}
	//cluster.Keyspace = "example"
	cluster.Consistency = gocql.Quorum
	cluster.ProtoVersion = 4
	cluster.ConnectTimeout = time.Second * 10
	session, err := cluster.CreateSession()
	if err != nil {
		panic(err)
	}
	println("connected")
	defer session.Close()

}`

**Comment by slayerjain** [comment]
In my case, we're able to connect via cli, but getting error if using the SDK.

**Comment by martin-sucha** [comment]
> In my case, we're able to connect via cli, but getting error if using the SDK.

cqlsh connects to a single node, gocql connects to all nodes in the cluster (but opens a control conn to one of the initially listed nodes first).

If one of the nodes returns error when connecting using protocol v4, cqlsh might either just connect to different node (depending on how you call it) or fallback to protocol v3. If you have ProtoVersion=4, gocql will only use v4.

So if there is issue connecting to just one node of the cluster and the control connection succeeds, the errors are just logged. If you provide gocql just one address to a node that shows the symptoms, the control connection fails and the session is not established (the `gocql: unable to dial control conn` error above).

I tried to reproduce with a single node 3.11.7 cluster, but couldn't reproduce.

If you have a single node where you can connect with cqlsh, but not with gocql, we could compare the TCP streams between cqlsh and gocql (e.g. to see whether cqlsh falls back to protocol version 3) if you attach network capture (and scrub the username/password of course or use a temporary one). Also if you connect to that node where you can't connect with gocql with cqlsh, which protocol version does it print?

Are all nodes in the cluster healthy?

**Comment by slayerjain** [comment]
@martin-sucha Yes, that makes sense. It turns out our nodes were connecting to nodes of other clusters too, and that could be causing the issue. https://github.com/k8ssandra/k8ssandra/issues/503

**Comment by martin-sucha** [comment]
Good. Thanks for information!

@behzadsh @abhi93104 does the above info help you solve your issues? Can we close this issue?

**Comment by OleksiienkoMykyta** [comment]
As the topic discussed and also has a related solved issue(https://github.com/apache/cassandra-gocql-driver/issues/1537), it can be closed.
But not sure about the status, is it stale or solved? The issue reporter hasn't responded since Feb 27, 2021.
Anyway, the candidate is to be closed.
